### PR TITLE
AST rewrite

### PR DESCRIPTION
**Pull request by bcardarella**
3x performance improvement in tests against dockyard.com

### Tasks
- [x] Replace binaries with AST
- [x] Pre-compile template AST and store in ETS
- [x] Store routes in ETS - paths and metadata (page_id, layout_id, and modules)
- [x] Lookup ETS routes by path (including paths with dynamic segments)
- [x] Render templates dynamically from AST stored in ETS
- [x] `@beacon_path_params`
- [x] Propagate updates
- [x] Double check docs and generator (new apps)

### CODE REVIEW

**Comment by bcardarella** [comment]
I've got to fix the test suite, and there are some additonal performance optimizations that can be gotten by reducing the number of functions generated.

File: lib/beacon/application.ex
Comment (bitwalker) [inline lib/beacon/application.ex:17]: Unless you plan to write to this table a lot, I would remove `write_concurrency` from this list. My assumption is that this is a table vastly dominated by reads. Without the option, you can of course still write to the table concurrently, it is just more expensive to do so.

File: lib/beacon/loader/db_loader.ex
Comment (bitwalker) [inline lib/beacon/loader/db_loader.ex:23]: Might be better to emit telemetry events instead of logging this (or emit logs at `debug` level instead). I know with my various open source projects people have often requested control over what is logged and when, and `telemetry` really provides the best degree of control over this (as well as having other benefits).

File: lib/beacon/loader/module_loader.ex
Comment (bitwalker) [inline lib/beacon/loader/module_loader.ex:3]: I don't think you need (nor should use) either `:code.delete/1` or `:code.purge/1` here. If there are ongoing requests to the server, this will result in those requests failing badly if they happen to be requesting something which references the affected code. You should leave this behavior to the ERTS code loader itself, which will ensure that ongoing requests complete using the previous version of the code, and any new requests use only the new code.

**Review by bitwalker — APPROVED** [review]
I left a few suggestions, but I'll leave it up to you on whether you want to use them. Looks good otherwise!

File: lib/beacon/loader/module_loader.ex
Comment (bcardarella) [reply lib/beacon/loader/module_loader.ex:3]: this would only get kicked off once on a page CRUD event. But with the move towards storing compiled pages in ets the need for this at all should go away

File: lib/beacon/application.ex
Comment (leandrocp) [reply lib/beacon/application.ex:17]: It won't be written very often. I'm gonna remove that opt.

File: lib/beacon/loader/db_loader.ex
Comment (leandrocp) [reply lib/beacon/loader/db_loader.ex:23]: Yep that's planned. In this case it was a quick mechanism for debugging but it'll be removed before merging 👍🏻

**Review by TheFirstAvenger — APPROVED** [review]

**Comment by bcardarella** [comment]
🎉

**Pull request merged**
